r/src/arrow_cpp11.h

```diff
@@ -138,6 +138,11 @@ inline R_xlen_t r_string_size(SEXP s) {
 }  // namespace unsafe
 
 inline SEXP utf8_strings(SEXP x) {
+  // ensure that x is not actually altrep first
+  if (ALTREP(x)) {
+    x = PROTECT(Rf_duplicate(x));
+    UNPROTECT(1);
```

Comment:
I'm pretty sure this isn't right, doesn't the `UNPROTECT` need to go after you're done using the thing you `PROTECT`ed?

I'm also not sure how the C-level PROTECT stuff interacts with unwind_protect.

Reply:
Hmmm, I'm surprised I don't get any stack imbalance issues with it. I _do_ get them if I don't have `UNPROTECT` or I call it unconditionally. But anyway I have a slightly different version incoming.